### PR TITLE
Implement the new PUT /agents/reconnect endpoint to force and agent reconnection to the manager

### PR DESCRIPTION
**Pull request by mcarmona99**
|Related issue|
|---|
| #7896 |

This PR closes https://github.com/wazuh/wazuh/issues/7896.

In this pull request, I have added the new `PUT /agents/reconnect` endpoint and its rbac action. I have been testing the endpoint and it works, as it can be seen in the following logs:

**MASTER:**
```
2021/03/19 16:08:35 wazuh-remoted[1994] ar-forward.c:40 at AR_Forward(): DEBUG: Active response request received: (msg_to_agent) [] N!S 001 force_reconnect
2021/03/19 16:08:35 wazuh-remoted[1994] ar-forward.c:99 at AR_Forward(): DEBUG: Active response sent: #!-force_reconnect
```

**AGENT:**
```
2021/03/19 16:09:13 wazuh-agentd[1673] receiver.c:92 at receive_msg(): DEBUG: Received message: '#!-force_reconnect'
2021/03/19 16:09:13 wazuh-agentd[1673] receiver.c:131 at receive_msg(): INFO: Wazuh Agent will be reconnected because a reconnect message was received
```

I have refactored `wazuh_queue.py`. I have created a function to create the `wazuh_queue` socket message.

I have updated the unit tests. In these tests, I have updated the wazuh_queue ones due to the refactor. I have also added new tests cases to cover the new endpoint's functions.

```
pytest framework/wazuh/tests/
============================== 744 passed, 12 warnings in 56.60s ===============================

pytest framework/wazuh/core/tests/
================================ 657 passed, 11 warnings in 2.43s ================================
```

I have updated the API integration tests. Test results:

```
test_agent_DELETE_endpoints.tavern.yaml 
	 6 passed, 8 warnings

test_agent_GET_endpoints.tavern.yaml 
	 91 passed, 98 warnings

test_agent_POST_endpoints.tavern.yaml 
	 5 passed, 7 warnings
```

New test cases added to (known test failing in `rbac_white_all`):

```
test_agent_PUT_endpoints.tavern.yaml
	 10 passed, 12 warnings

test_rbac_black_agent_endpoints.tavern.yaml 
	 41 passed, 43 warnings

test_rbac_white_agent_endpoints.tavern.yaml 
	 41 passed, 43 warnings

test_rbac_white_all_endpoints.tavern.yaml 
	 1 failed, 148 passed, 3 xfailed, 154 warnings
```

Regards,
Manuel.

### CODE REVIEW

**Review by Selutario — CHANGES_REQUESTED** [review]
Everything looks correct. Rebase with the core branch and solve conflicts.

**Review by Selutario — APPROVED** [review]

**Review by davidjiglesias — CHANGES_REQUESTED** [review]

**Comment by mcarmona99** [comment]
Manual tests and more changes:

`https://localhost:55050/security/policies?policy_ids=2
`

**4.1.4:**
```
{
  "data": {
    "affected_items": [
      {
        "id": 2,
        "name": "agents_all_agents",
        "policy": {
          "actions": [
            "agent:read",
            "agent:delete",
            "agent:modify_group",
            "agent:restart",
            "agent:upgrade"
          ],
          "resources": [
            "agent:id:*",
            "agent:group:*"
          ],
          "effect": "allow"
        },
        "roles": [
          1,
          5
        ]
      }
    ],
    "total_affected_items": 1,
    "total_failed_items": 0,
    "failed_items": []
  },
  "message": "All specified policies were returned",
  "error": 0
}
```

**Upgrade to 4.2.0 + reconnect:**
```
{
  "data": {
    "affected_items": [
      {
        "id": 2,
        "name": "agents_all_agents",
        "policy": {
          "actions": [
            "agent:read",
            "agent:delete",
            "agent:modify_group",
            "agent:reconnect",
            "agent:restart",
            "agent:upgrade"
          ],
          "resources": [
            "agent:id:*",
            "agent:group:*"
          ],
          "effect": "allow"
        },
        "roles": [
          1,
          5
        ]
      }
    ],
    "total_affected_items": 1,
    "total_failed_items": 0,
    "failed_items": []
  },
  "message": "All specified policies were returned",
  "error": 0
}
```

File: api/api/spec/spec.yaml [hunk 1]
```diff
@@ -16,7 +16,7 @@ info:
 
     Login with USER and PASSWORD:
 
-    `curl -u <USER>:<PASSWORD> -k -X GET "https://<HOST_IP>:55000/security/user/authenticate`
+    `curl -u <USER>:<PASSWORD> -k -X GET "https://<HOST_IP>:55000/security/user/authenticate"`
```
[the remark this answers is not in the document]
Comment (mcarmona99) [reply api/api/spec/spec.yaml:19]: This change is already in master

**Comment by mcarmona99** [comment]
### New test results:

```
framework/wazuh/core/tests
================== 657 passed, 9 warnings in 2.03s ==================

framework/wazuh/tests/
================= 759 passed, 11 warnings in 52.67s =================
```
```

test_agent_PUT_endpoints.tavern.yaml 
	 10 passed, 12 warnings
```

**Comment by mcarmona99** [comment]
Manual tests to `wazuh_queue` after the refactor.

### Syscheck:

**MANAGER (api.log)**

`2021/03/29 08:35:23 INFO: wazuh 172.21.0.1 "PUT /syscheck" with parameters {"agents_list": "001"} and body {} done in 0.066s: 200`

**AGENT (ossec.log)**

```
2021/03/29 08:34:12 wazuh-syscheckd: INFO: (6008): File integrity monitoring scan started.
2021/03/29 08:34:12 rootcheck: INFO: Starting rootcheck scan.
2021/03/29 08:34:12 wazuh-syscheckd: INFO: (6009): File integrity monitoring scan ended.
2021/03/29 08:34:21 rootcheck: INFO: Ending rootcheck scan.
```


### Reconnect

**MANAGER (api.log)**

`2021/03/29 08:38:04 INFO: wazuh 172.21.0.1 "PUT /agents/reconnect" with parameters {"agents_list": "001"} and body {} done in 0.068s: 200`

**AGENT (ossec.log)**

```
2021/03/29 08:38:04 wazuh-agentd: INFO: Wazuh Agent will be reconnected because a reconnect message was received
2021/03/29 08:38:04 wazuh-agentd: INFO: Closing connection to server (nginx-lb/172.21.0.2:1514/tcp).
2021/03/29 08:38:04 wazuh-agentd: INFO: Trying to connect to server (nginx-lb/172.21.0.2:1514/tcp).
2021/03/29 08:38:04 wazuh-agentd: INFO: (4102): Connected to the server (nginx-lb/172.21.0.2:1514/tcp).
```


### Restart

**MANAGER (api.log)**

`2021/03/29 08:39:18 INFO: wazuh 172.21.0.1 "PUT /agents/001/restart" with parameters {} and body {} done in 0.101s: 200
`

**AGENT (ossec.log)**

```
2021/03/29 08:39:18 wazuh-modulesd:syscollector: INFO: Destroy received for Syscollector.
2021/03/29 08:39:18 wazuh-modulesd:syscollector: INFO: Module finished.
2021/03/29 08:39:18 wazuh-logcollector: INFO: (1225): SIGNAL [(15)-(Terminated)] Received. Exit Cleaning...
2021/03/29 08:39:18 wazuh-syscheckd: INFO: (1225): SIGNAL [(15)-(Terminated)] Received. Exit Cleaning...
2021/03/29 08:39:19 wazuh-agentd: INFO: (1225): SIGNAL [(15)-(Terminated)] Received. Exit Cleaning...
2021/03/29 08:39:19 wazuh-execd: INFO: (1314): Shutdown received. Deleting responses.
2021/03/29 08:39:19 wazuh-execd: INFO: (1225): SIGNAL [(15)-(Terminated)] Received. Exit Cleaning...
2021/03/29 08:39:20 wazuh-execd: INFO: Started (pid: 3115).
2021/03/29 08:39:21 wazuh-agentd: INFO: (1410): Reading authentication keys file.
2021/03/29 08:39:21 wazuh-agentd: INFO: Using notify time: 10 and max time to reconnect: 60
...
(we can also see new reconnect module info in logs)
```

### Restart via active response:

**MANAGER (api.log)**

`2021/03/29 08:42:39 INFO: wazuh 172.21.0.1 "PUT /active-response" with parameters {"agents_list": "001"} and body {"command": "restart-wazuh"} done in 0.081s: 200`

**AGENT (ossec.log)**

`Same as above`

Regards,
Manuel.

**Review by davidjiglesias — CHANGES_REQUESTED** [review]
Review exception raise in wazuhqueu send_msg_to_agent and active-response disabled exception

**Review by davidjiglesias — APPROVED** [review]

**Pull request merged**